Add new parameters to manage mtime and atime for file module, fixes #30226

StephenSorriaux opened the pull request:
##### SUMMARY
I added a way to manage mtime and atime of files by adding 4 new parameters.

Fixes #30226
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
file

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (feat/new-parameter-for-file-module 7c3669c10f) last updated 2018/07/24 22:46:46 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/stephen/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/stephen/dev/ansible/lib/ansible
  executable location = /home/stephen/dev/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
The 4 new parameters:

- ```modification_time```: controls the mtime of the file. Can be 'preserve' to keep the current mtime, 'now' to use the current timestamp or a string that follows time format defined by modification_time_format. Default is ```None``` which keeps the same behaviour as now (no changes for ```state=[file,link,hard,directory]``` and always change for ```state=touch```) ;

- ```modification_time_format```: add a way to define a custom format, default is as suggested (YYYYmmddHHMM.SS) ;

- ```access_time```: same as modification_time but for atime ;

- ```access_time_format```: same as modification_time_format but for atime.

Each times are given using local timezone but it can be changed using a custom time format.

Those parameters are compatible with every states (file, touch, directory, link, hard).

```state=touch``` is now idempotent using ```modification_time=preserve``` and ```access_time=preserve``` :)

Regards,
Stephen

StephenSorriaux committed:
Add new parameters related to mtime and atime for file module, fixes #30226

abadger commented:
I think it would be better to add parameters to the module to set the timestamp.  If not given, the timestamp will be set to now (Unfortunately, needed for backwards compatibility :-(  ) If given, the timestampe will be set to that value.  We can probably use a special value to men don't change the timestamp as well.

We probably need parameters for both atime and mtime to do this.

I'm not sure if we should support it for the other subcommands, not just touch (timestamps are similar to permissions in that they're really just file metadata.)

I just read the GNU touch info page and am left a bit unnerved by the range of ways to specify a timestamp to that version of touch.  I'd suggest to keep it simple, perhaps a simplification of the -t format: ```YYYYMMDDHHMM[.SS]```.  I'm not sure whether to make the timestamp use the local timezone as -t does or make it always use UTC.

StephenSorriaux commented:
Sure, I can add those features. Don't you think it could be better to create a new module "touch" to handle all of this?

abadger commented:
If I had a time machine.... ;-).

bcoca commented:
+10k for touch module ... I had prototype somewhere ...

abadger commented:
To expand, we currently have the tuoch subcommand of file and we can't go backwards in time.  So we have to fix it.  And adding these things to the touch subcommand is the way to fix it.

abadger commented:
I've been thinking about one of the open questions, whether setting timestamp needs to be supported for the other subcommands and I think the answer is yes.

All the subcommands support setting permissions so I think this falls in the same category.  This may be an opportunity for us to change the default of the timestamp setting for touch as well (to not setting it) as none of the other subcommands will want to set it by default (which will make touch idempotent by default)

StephenSorriaux commented:
I just updated the PR following your advices by adding 4 new parameters and removing the first 2 I was submitting.

So, concerning the changes, I added:

- modification_time: controls the mtime of the file. Can be 'same' (which is default) to keep the current mtime, 'now' to use the current timestamp or a string that follows time format defined by modification_time_format ;
- modification_time_format: add a way to define a custom format, default is as suggested (YYYYmmddHHMM.SS) ;
- access_time: same as modification_time but for atime ;
- access_time_format: same as modification_time_format but for atime.

Each times are given using local timezone but it can be changed using a custom time format.

Those parameters are compatible with every states (file, touch, directory, link, hard).

By default, state=touch is now idempotent :)

I first wanted to add the logic inside the module.set_fs_attributes_if_different() method but since it is used by several modules, I did not want to interfer in any way.

jborean93 commented on lib/ansible/modules/files/file.py:
```diff
@@ -49,7 +49,6 @@
         If C(touch) (new in 1.4), an empty file will be created if the C(path) does not
         exist, while an existing file or directory will receive updated file access and
         modification times (similar to the way `touch` works from the command line).
-    default: file
```
While the argspec does not specify file as the default, it is the default if `path` does not exist. Not sure if we should remove it, if we do remove it then this should be added to the description of state

StephenSorriaux replied:
Yes you're right, I think I will simply rollback this since the case was already managed by test/sanity/validate-modules/ignore.txt

StephenSorriaux replied:
rollbacked

jborean93 commented on lib/ansible/modules/files/file.py:
```diff
@@ -76,6 +75,30 @@
+    description:
+      - This parameter indicates to which time should the file's modification time be set
+      - 'Should be C(same) (default) when no modification is required, C(YYYYMMDDHHMM.SS) when using default time format, or C(now)'
+    default: same
```
Should the default be `now` instead as that's what is the current behaviour when `state: touch`?

StephenSorriaux replied:
Since those parameters are used for every states available, if I change the default to ```now```, we will change the current behaviour of all states (except ```state: touch```) and loose their idempotence

jborean93 replied:
I see what you mean, maybe not having it defined (by default) keeps the same behaviour where touch will always update and the other states won't update. If a user explicitly defines, `now`, `same`, or `<timestamp>` then we can set accordingly.

abadger replied:
If we can get away with it, I'd like to change this to same so that touch is idempotent by default.  But probably need to run it by the other committers.  It is backwards incompatible but more sensible API

(If the other committers deem backwards compatibility to be that important, then jborean's proposal sounds best.  If timestamp is None: use correct default for the state specified)

StephenSorriaux replied:
Sure, awaiting more reviews then :)

abadger replied:
At yesterday's meeting we didn't quite decide the question of the default value decisively (only three people registered an opinion on it) but of the three there seemed to be more enthusiasm for adding the ability for "state=file" to create files (phased in via a deprecation period because of backwards compatibility concerns) rather than having touch default to keeping the same timestamp.  So if you want to be proactive, you should probably update this to have the split defaults as @jborean93 suggested.

There was some bikeshedding about the special names as well:
* There was a desire to make None mean "same" but I don't think we can do that with the split default (None means, do the default.  And since it has two meanings, we need a different special word to mean keep the timestamp the same)
* In the end people thought that "preserve" would be better than "same" (since that's used in copy to keep a file source's mode which is somewhat similar.)

StephenSorriaux commented:
Hum... I was not aware that amending a commit would remove the comments posted here...

So I updated the PR following @jborean93 advices:

- default for acces_time and modification_time are None, which keeps the same behaviour for all states
- ```same``` is changed in favour of ```preserve```

I also added 2 tests.

If possible, I would be happy to help updating this module and changing the current ```state=file``` behaviour :+1:

abadger committed:
Just fix some typos

StephenSorriaux commented:
Thanks a lot!

OK, I will submit a PR for this new behaviour.
